crates/turbopack-ecmascript/src/references/mod.rs

```diff
@@ -1329,55 +1355,83 @@ pub(crate) async fn analyze_ecmascript_module(
-                                                state.origin.context(),
-                                                context,
-                                            )
-                                            .into()
-                                        }),
-                                        RequestVc::parse(Value::new(request.clone().into())),
-                                        Default::default(),
-                                        state
-                                            .import_parts
-                                            .then(|| {
-                                                export.as_ref().map(|export| {
-                                                    ModulePartVc::export(export.to_string())
-                                                })
-                                            })
-                                            .flatten(),
-                                    )
-                                    .resolve()
-                                    .await?;
-                                    analysis.add_reference(esm_reference);
-                                    analysis.add_code_gen(EsmBindingVc::new(
-                                        esm_reference,
-                                        export.clone(),
-                                        AstPathVc::cell(ast_path.to_vec()),
-                                    ));
-                                }
+                            if handle_free_var_reference(ast_path, value, state, analysis).await? {
+                                return Ok(());
                             }
-                            break;
                         }
                     }
                 }
 
                 Ok(())
             }
 
+            async fn handle_free_var_reference(
+                ast_path: &[AstParentKind],
+                value: &FreeVarReference,
+                state: &AnalysisState<'_>,
+                analysis: &mut AnalyzeEcmascriptModuleResultBuilder,
+            ) -> Result<bool> {
+                // We don't want to replace assignments as this would lead to invalid code.
+                if ast_path.len() > 4
+                    && matches!(
+                        ast_path[ast_path.len() - 1],
+                        AstParentKind::Expr(ExprField::Member)
+                    )
+                    && matches!(
+                        ast_path[ast_path.len() - 2],
+                        AstParentKind::Pat(PatField::Expr)
+                    )
+                    && matches!(
+                        ast_path[ast_path.len() - 3],
+                        AstParentKind::PatOrExpr(PatOrExprField::Pat)
+                    )
+                    && matches!(
+                        ast_path[ast_path.len() - 4],
+                        AstParentKind::AssignExpr(AssignExprField::Left)
+                    )
+                {
```

Comment:
Rust matches are pretty powerful:

```suggestion
                if matches!(
                    &ast_path[..],
                    [
                        ..,
                        _,
                        AstParentKind::AssignExpr(AssignExprField::Left),
                        AstParentKind::PatOrExpr(PatOrExprField::Pat),
                        AstParentKind::Pat(PatField::Expr),
                        AstParentKind::Expr(ExprField::Member),
                    ]
                ) {
```